librz/asm/arch/sh/disassembler.c

```diff
@@ -0,0 +1,471 @@
+	{ "stc.l", SH_OP_STC, OPCODE(4, M, 2, 3), 0x0f00, SH_SCALING_L, { PARAM(VBR, SH_REG_DIRECT), ADDR(NIB2, SH_REG_INDIRECT_D) } },
+	{ "stc.l", SH_OP_STC, OPCODE(4, M, 3, 3), 0x0f00, SH_SCALING_L, { PARAM(SSR, SH_REG_DIRECT), ADDR(NIB2, SH_REG_INDIRECT_D) } },
+	{ "stc.l", SH_OP_STC, OPCODE(4, M, 4, 3), 0x0f00, SH_SCALING_L, { PARAM(SPC, SH_REG_DIRECT), ADDR(NIB2, SH_REG_INDIRECT_D) } },
+	{ "stc.l", SH_OP_STC, OPCODE(4, M, 3, 2), 0x0f00, SH_SCALING_L, { PARAM(SGR, SH_REG_DIRECT), ADDR(NIB2, SH_REG_INDIRECT_D) } },
+	{ "stc.l", SH_OP_STC, OPCODE(4, M, f, 2), 0x0f00, SH_SCALING_L, { PARAM(DBR, SH_REG_DIRECT), ADDR(NIB2, SH_REG_INDIRECT_D) } },
+	{ "stc.l", SH_OP_UNIMPL, OPCODE(4, N, M, 3), 0x0f70, SH_SCALING_L, { ADDR(NIB1, SH_REG_DIRECT), ADDR(NIB2, SH_REG_INDIRECT_D) } },
+	{ "sts", SH_OP_STS, OPCODE(0, N, 0, a), 0x0f00, SH_SCALING_INVALID, { PARAM(MACH, SH_REG_DIRECT), ADDR(NIB2, SH_REG_DIRECT) } },
+	{ "sts", SH_OP_STS, OPCODE(0, N, 1, a), 0x0f00, SH_SCALING_INVALID, { PARAM(MACL, SH_REG_DIRECT), ADDR(NIB2, SH_REG_DIRECT) } },
+	{ "sts", SH_OP_STS, OPCODE(0, N, 2, a), 0x0f00, SH_SCALING_INVALID, { PARAM(PR, SH_REG_DIRECT), ADDR(NIB2, SH_REG_DIRECT) } },
+	{ "sts.l", SH_OP_STS, OPCODE(4, N, 0, 2), 0x0f00, SH_SCALING_L, { PARAM(MACH, SH_REG_DIRECT), ADDR(NIB2, SH_REG_INDIRECT_D) } },
+	{ "sts.l", SH_OP_STS, OPCODE(4, N, 1, 2), 0x0f00, SH_SCALING_L, { PARAM(MACL, SH_REG_DIRECT), ADDR(NIB2, SH_REG_INDIRECT_D) } },
+	{ "sts.l", SH_OP_STS, OPCODE(4, N, 2, 2), 0x0f00, SH_SCALING_L, { PARAM(PR, SH_REG_DIRECT), ADDR(NIB2, SH_REG_INDIRECT_D) } },
+	{ "trapa", SH_OP_UNIMPL, OPCODE(c, 3, I, I), 0x00ff, SH_SCALING_INVALID, { ADDR(NIB0, SH_IMM_U), NOPARAM } }
+
+	/* floating-point single-precision instructions */
+	// TODO
+};
+
+#undef NOPARAM
+#undef PARAM
+#undef ADDR
+#undef OPCODE
+#undef OPCODE_
+#undef M
+#undef D
+#undef N
+#undef I
+
+/**
+ * \brief Disassemble \p opcode and return a SHOp
+ *
+ * \param opcode 16 bit wide opcode
+ * \return SHOp object corresponding to the opcode
+ */
+RZ_API RZ_OWN SHOp *sh_disassembler(ut16 opcode) {
+	ut32 opcode_num = sizeof(sh_op_lookup) / sizeof(SHOpRaw);
+
+	for (ut16 i = 0; i < opcode_num; i++) {
+		if ((opcode | sh_op_lookup[i].mask) == sh_op_lookup[i].opcode) {
+			SHOpRaw raw = sh_op_lookup[i];
+			SHOp *op = RZ_NEW(SHOp);
+			op->opcode = opcode;
+			op->mnemonic = raw.mnemonic;
+			op->scaling = raw.scaling;
+			op->str_mnem = raw.str_mnem;
+			// check for "weird" mov.l
+			if (raw.opcode == MOVL) {
+				op->param[0] = sh_op_get_param_movl(opcode, true);
+				op->param[1] = sh_op_get_param_movl(opcode, false);
+				return op;
+			}
+			op->param[0] = sh_op_get_param(opcode, raw.param_builder[0]);
+			op->param[1] = sh_op_get_param(opcode, raw.param_builder[1]);
+			return op;
+		}
+	}
+
+	RZ_LOG_WARN("SuperH: Invalid opcode encountered by disassembler")
+	return NULL;
+}
+
+#define MNEM_PADLEN 10
+#define ARG_PADLEN  14
+
+RZ_API RZ_OWN char *sh_op_param_to_str(SHParam param) {
```

Comment:
Doxygen? Also, I am not really sure it should be the API function.

Reply:
I think it will be good to have an API function for disassembly.